// framer-motion/dist/cli.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = void 0;
const exporter_js_1 = require("./exporter.js");
const json5_1 = __importDefault(require("json5"));
const events_1 = require("events");
const chokidar_1 = __importDefault(require("chokidar"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const find_up_1 = __importDefault(require("find-up"));
const path_1 = __importStar(require("path"));
const configNames = ['unframer.config.json', 'unframer.json'];
const cac_1 = require("cac");
const utils_js_1 = require("./utils.js");
exports.cli = (0, cac_1.cac)('unframer');
exports.cli.command('', 'Run unframer')
    .option('--watch', 'Watch for Framer and unframer.config.json changes')
    .action(function main(options) {
    return __awaiter(this, void 0, void 0, function* () {
        fixOldUnframerPath();
        const cwd = process.cwd();
        const watch = process.argv.includes('--watch');
        utils_js_1.logger.log(`Looking for ${configNames.join(', ')} in ${cwd}`);
        const configPath = yield (0, find_up_1.default)(configNames, { cwd });
        if (!configPath) {
            utils_js_1.logger.log(`No ${configNames.join(', ')} found`);
            return;
        }
        let configBasename = (0, path_1.basename)(configPath);
        const configContent = fs_extra_1.default.readFileSync(configPath, 'utf8');
        if (!configContent) {
            utils_js_1.logger.log(`No ${configBasename} contents found`);
            return;
        }
        let config = json5_1.default.parse(configContent);
        let controller = new AbortController();
        (0, events_1.setMaxListeners)(0, controller.signal);
        processConfig({
            config,
            watch,
            signal: controller.signal,
            configBasename,
        });
        if (!watch) {
            return;
        }
        const watcher = chokidar_1.default.watch(configPath, {
            persistent: true,
        });
        watcher.on('change', (path) => __awaiter(this, void 0, void 0, function* () {
            utils_js_1.logger.log(`${configBasename} changed`);
            console.log();
            controller.abort();
            controller = new AbortController();
            (0, events_1.setMaxListeners)(0, controller.signal);
            const newConfig = safeJsonParse(fs_extra_1.default.readFileSync(configPath, 'utf8'));
            if (!newConfig) {
                utils_js_1.logger.log(`Invalid ${configBasename} file`);
                return;
            }
            const newNames = getNewNames(config, newConfig);
            if (newNames.length) {
                utils_js_1.logger.log(`New components found: ${newNames.join(', ')}`);
                yield processConfig({
                    config: Object.assign(Object.assign({}, newConfig), { components: pluck(newConfig.components, newNames) }),
                    watch,
                    configBasename,
                    // signal: controller.signal,
                });
            }
            config = newConfig;
        }));
    });
});
const defaultConfig = `{
    "schema": "https://unframer-schema.vercel.app/schema.json",
    "outDir": "./framer",
    "components": {
        // add here your Framer components urls, the code will be written to outDir/{componentName}.js
        "example-hero": "https://framer.com/m/Header-WtSW.js",
    }
}
`;
function fixOldUnframerPath() {
    // if unframer.json exists, rename it to unframer.config.json
    const oldConfigPath = fs_extra_1.default.existsSync('unframer.json');
    if (oldConfigPath) {
        fs_extra_1.default.renameSync('unframer.json', 'unframer.config.json');
        utils_js_1.logger.green('legacy unframer.json config renamed to unframer.config.json');
        return true;
    }
    return false;
}
exports.cli.command('init', 'Init the unframer.config.json config').action((options) => __awaiter(void 0, void 0, void 0, function* () {
    let fixed = fixOldUnframerPath();
    if (fixed) {
        return;
    }
    fs_extra_1.default.writeFileSync(`unframer.config.json`, defaultConfig);
    const p = path_1.default.resolve(process.cwd(), 'unframer.config.json');
    console.log(`${p} file created`);
}));
const version = require('../package.json').version;
exports.cli.version(version).help();
function safeJsonParse(json) {
    try {
        return json5_1.default.parse(json);
    }
    catch (e) {
        return null;
    }
}
function pluck(o, names) {
    return Object.fromEntries(names.map((n) => [n, o[n]]));
}
function getNewNames(oldConfig, newConfig) {
    // get the new names, also check if the previous url (object value) has changed
    const oldKeys = Object.keys(oldConfig.components);
    const newKeys = Object.keys(newConfig.components);
    const newNames = newKeys.filter((key) => {
        if (!oldKeys.includes(key)) {
            return true;
        }
        if (oldConfig.components[key] !== newConfig.components[key]) {
            return true;
        }
        return false;
    });
    return newNames;
}
function processConfig(_a) {
    return __awaiter(this, arguments, void 0, function* ({ config, watch, signal, configBasename, }) {
        try {
            const { components, breakpoints, outDir } = config || {};
            const installDir = path_1.default.resolve(process.cwd(), outDir || 'framer');
            if (!components) {
                utils_js_1.logger.log(`No components found in ${configBasename}`);
                return;
            }
            yield (0, exporter_js_1.bundle)({
                components,
                breakpoints,
                cwd: installDir,
                watch,
                signal,
            });
        }
        catch (e) {
            if (signal) {
                utils_js_1.logger.log('Error processing config', e.stack);
                return;
            }
            throw e;
        }
    });
}
